### PR TITLE
feat(ios): configurable keyboard height

### LINKED ISSUE #5897: feat(ios): Touch keyboard height adjustability
Described for Android in #3225. We need a similar fix for iOS.

---

**Keyman for iPhone/iPad/Android:**
 - Device: iPhone/iPad 
 - OS: iOS 
 - Keyman version: 15.0

### PR DESCRIPTION
**Pull request by sgschantz**
Allows the height of the Keyman keyboard to be adjusted by the user from within the Keyman Settings.

Fixes #5897

# User Testing

* **TEST_PORTRAIT_KEYBOARD_HEIGHT_ADJUSTMENT**:

1. Install and run the Keyman build associated with this PR
1. Go to Keyman Settings with the Keyman app
1. Holding the device in portrait orientation, tap Adjust Keyboard Height
1. Drag the resizer arrow to adjust the height of the keyboard
1. Tap the Back button
1. Tap the Done button
1. Confirm that the keyboard has been resized within the Keyman app

* **TEST_LANDSCAPE_KEYBOARD_HEIGHT_ADJUSTMENT**:

1. Install and run the Keyman build associated with this PR
1. Go to Keyman Settings with the Keyman app
1. Holding the device in landscape orientation, tap Adjust Keyboard Height
1. Drag the resizer arrow to adjust the height of the keyboard
1. Tap the Back button
1. Tap the Done button
1. Confirm that the keyboard has been resized within the Keyman app

* **TEST_SYSTEM_KEYBOARD_HEIGHT_ADJUSTMENT**:

1. Set the keyboard height to a size noticeably different than the default size
1. Go to an iOS application with text input
1. Switch to the Keyman system keyboard and confirm that the keyboard height is the same as it was in Keyman

* **TEST_ADJUST_KEYBOARD_ROTATION**:

1. Go to Keyman Settings within the Keyman app
1. Tap Adjust Keyboard Height
1. Rotate the device between portrait and landscape
1. Confirm that the animation is smooth and the keyboard size changes correctly after rotating

* **TEST_ADJUST_KEYBOARD_DARK_MODE**:

1. Go to Keyman Settings within the Keyman app
1. Tap Adjust Keyboard Height
1. Go to iOS Settings > Display & Brightness
1. Change Appearance from Light to Dark or vice-versa
1. Confirm that the blank keyboard displayed on the Adjust Keyboard screen reflect the appearance change

* **TEST_RESET_TO_DEFAULT**:

1. Go to Keyman Settings within the Keyman app
1. Tap Adjust Keyboard Height
1. Resize the keyboard to a height noticeably different from the default
1. Tap the 'Rest to Default Keyboard Height' button
1. Confirm that they keyboard size springs back to its original default size

### CODE REVIEW

File: ios/engine/KMEI/KeymanEngine/Classes/KeyboardHeightViewController.swift [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+  private func determineOrientation() {
+    self.isPortrait = UIScreen.main.bounds.height > UIScreen.main.bounds.width
+    let message = "determineOrientation, isPortrait: \(self.isPortrait)"
+    os_log("%{public}s", log:KeymanEngineLogger.ui, type: .debug, message)
+  }
+  
+  private func determineDefaultKeyboardHeights() {
+    // if no KeyboardScaleMap found for device, then default to 216.0
+    let portraitKeyboardScale = KeyboardScaleMap.getDeviceDefaultKeyboardScale(forPortrait: true)
+    self.defaultPortraitHeight = Double(portraitKeyboardScale?.keyboardHeight ?? 216.0)
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngine/Classes/KeyboardHeightViewController.swift:72]: 216.0 appears in a few places. Is it worth making it a constant somewhere?

File: ios/engine/KMEI/KeymanEngine/Classes/KeyboardHeightViewController.swift [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+      if (Storage.active.userDefaults.object(forKey: Key.portraitKeyboardHeight) != nil) {
+        self.keyboardHeight = Storage.active.userDefaults.landscapeKeyboardHeight
+        let message = "applyKeyboardHeight, from UserDefaults loaded landscape value \(self.keyboardHeight)"
+        os_log("%{public}s", log:KeymanEngineLogger.ui, type: .debug, message)
+      } else {
+        self.keyboardHeight = self.defaultLandscapeHeight
+        let message = "applyKeyboardHeight, landscapeHeight not found in UserDefaults, using default value \(self.keyboardHeight)"
+        os_log("%{public}s", log:KeymanEngineLogger.ui, type: .error, message)
+      }
+    }
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngine/Classes/KeyboardHeightViewController.swift:104]: There's a fair bit of repetition between landscape and portrait orientation code. Could we merge the values into an array (0=portait, 1=landscape) or something like that? What do you think? Or an {x,y}?

**Review by mcdurdin — APPROVED** [review]
LGTM, just nits. Good work

**Comment by dinakaranr** [comment]
# Test Results
I tested this issue with the attached "Keyman 18.0.130-alpha-test-12571" build(test flight) on the iOS 17.4, iPhone 13 physical device.

* **TEST_PORTRAIT_KEYBOARD_HEIGHT_ADJUSTMENT (Passed):**
1. I installed the Keyman 18.0.130-alpha-test-12571 version.
2. Open the Keyman app. Go to the keyman settings.
3. iPhone device holds in portrait view. Tap on the "Adjust Keyboard Height" option.
4. Drag the resizer arrow to adjust the height of the keyboard(Small to Big Height).
5. Tap the Back button.
6. Tap the Done button.
In Keyman editor: Verified that the keyboard resized within the Keyman app.

* **TEST_LANDSCAPE_KEYBOARD_HEIGHT_ADJUSTMENT (Passed):**
1. I installed the Keyman 18.0.130-alpha-test-12571 version.
2. Open the Keyman app. Go to the keyman settings.
3. iPhone device holds in landscape view. Tap on the "Adjust Keyboard Height" option.
4. Drag the resizer arrow to adjust the height of the keyboard(Big Height to Small).
5. Tap the Back button
6. Tap the Done button
In Keyman editor: Verified that the keyboard resized within the Keyman app.

* **TEST_SYSTEM_KEYBOARD_HEIGHT_ADJUSTMENT (Passed):**
1. I installed the Keyman 18.0.130-alpha-test-12571 version.
2. Open the Keyman app. Go to the keyman settings.
3. iPhone device holds in portrait view. Tap on the "Adjust Keyboard Height" option.
4. Drag the resizer arrow to increase the small to Big height of the keyboard.
5. Tap the Back button
6. Tap the Done button
7. Set the keyman keyboard as a system keyboard by clicking keyman/Settings/System Keyboard Settings/Keyboards/Allow full access & Keyman
8. Open the Pages app. Open a blank page.
9. Click on the blank page.
10. In Keyman editor: Verified that the keyboard resized with bigger. 

* **TEST_ADJUST_KEYBOARD_ROTATION (Passed):**
1. I installed the Keyman 18.0.130-alpha-test-12571 version.
2. Open the Keyman app. Go to the keyman settings.
3. iPhone device holds in portrait view. Tap on the "Adjust Keyboard Height" option.
4. Rotate the device from portrait to landscape
5. Verified that the keyboard animation appeared as small size in the landscape view.
6. Rotate the device from landscape to portrait.
7. Verified that the keyboard height appeared as a bigger size in the portrait view.
8. Verified that the keyboard UI appeared correctly when rotating the view in the phone.

* **TEST_ADJUST_KEYBOARD_DARK_MODE (Passed):**
1. I installed the Keyman 18.0.130-alpha-test-12571 version.
2. Open the Keyman app. Go to the keyman settings.
3. iPhone device holds in portrait view. Tap on the "Adjust Keyboard Height" option.
4. Change the iPhone theme from light to dark mode.
5. Verified that the keyboard animation appeared in the dark view.
6. Verified that the white background appeared with blue & black text.
7. Verified that the blue text is clickable.

* **TEST_RESET_TO_DEFAULT (Passed):**
1. I installed the Keyman 18.0.130-alpha-test-12571 version.
2. Open the Keyman app. Go to the keyman settings.
3. iPhone device holds in portrait view. Tap on the "Adjust Keyboard Height" option.
4. Drag the resizer arrow to adjust the height of the keyboard(Small to Big Height).
5. Tap the Back button
6. Tap the Done button
7. In Keyman editor: Verified that the keyboard resized within the Keyman app.
8. Go to the keyman settings. Tap on the "Adjust Keyboard Height" option.
9. Click on the 'Rest to Default Keyboard Height' button.
10. Verified that the keyboard is reset to the default size on the keyboard animation.
11. Tap the Back button
12. Tap the Done button
13. In Keyman editor: Verified that the keyboard appeared as the default size in the Keyman app.

File: ios/engine/KMEI/KeymanEngine/Classes/KeyboardHeightViewController.swift (hunk 1 above)
Comment (sgschantz) [reply ios/engine/KMEI/KeymanEngine/Classes/KeyboardHeightViewController.swift:72]: I created a new constant `unknownDeviceKeyboardHeight`

File: ios/engine/KMEI/KeymanEngine/Classes/KeyboardHeightViewController.swift (hunk 2 above)
Comment (sgschantz) [reply ios/engine/KMEI/KeymanEngine/Classes/KeyboardHeightViewController.swift:104]: I tried to make the new code just use a keyboard height where I could and be unaware of orientation.  For the code that needs to be aware of orientation, I don't see any real advantage over using individual named fields. I think we can assume there will only be two, and we don't have mentally map to some other structure when reading or writing the values based on the current orientation.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.160-alpha](https://keyman.com/downloads/releases/alpha/18.0.160)

**Pull request merged**